cdc/processor/sinkmanager/redo_log_advancer.go

```diff
@@ -0,0 +1,265 @@
+// limitations under the License.
+
+package sinkmanager
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/memquota"
+	"github.com/pingcap/tiflow/cdc/processor/sourcemanager/engine"
+	"github.com/pingcap/tiflow/cdc/redo"
+	"go.uber.org/zap"
+)
+
+type redoLogAdvancer struct {
+	// NOTICE: This task is immutable, so please never modify it.
+	task *redoTask
+	// redoDMLManager is used to write the redo log.
+	redoDMLManager redo.DMLManager
+	// memQuota is used to acquire memory quota for the redo log writer.
+	memQuota *memquota.MemQuota
+	// NOTICE: First time to run the task, we have initialized memory quota for the table.
+	// It is defaultRequestMemSize.
+	availableMem uint64
+	// How much memory we have used.
+	// This is used to calculate how much memory we need to acquire.
+	// Only when usedMem > availableMem we need to acquire memory.
+	usedMem uint64
+	// Used to record the last written position.
+	// We need to use it to update the lower bound of the table sink.
+	lastPos engine.Position
+	// Buffer the events to be written to the redo log.
+	events []*model.RowChangedEvent
+
+	// emittedCommitTs is used to record the last emitted transaction commit ts.
+	emittedCommitTs uint64
+	// Used to record the latest written transaction commit ts.
+	lastTxnCommitTs uint64
+
+	// pendingTxnSize used to record the size of the uncommitted events.
+	pendingTxnSize uint64
+	// Used to record the current transaction commit ts.
+	currTxnCommitTs uint64
+}
+
+func newRedoLogAdvancer(
+	task *redoTask,
+	memQuota *memquota.MemQuota,
+	availableMem uint64,
+	redoDMLManager redo.DMLManager,
+) *redoLogAdvancer {
+	return &redoLogAdvancer{
+		task:           task,
+		memQuota:       memQuota,
+		availableMem:   availableMem,
+		events:         make([]*model.RowChangedEvent, 0, bufferSize),
+		redoDMLManager: redoDMLManager,
+	}
+}
+
+// advance tries to emit the events to the redo log manager and
+// advance the resolved ts of the redo log manager.
+func (a *redoLogAdvancer) advance(ctx context.Context, cachedSize uint64) error {
+	if len(a.events) > 0 {
+		// releaseMem is used to release the memory quota
+		// after the events are written to redo log.
+		// It more like a callback function.
+		var releaseMem func()
+		refundMem := a.pendingTxnSize - cachedSize
+		if refundMem > 0 {
+			releaseMem = func() {
+				a.memQuota.Refund(refundMem)
+				log.Debug("MemoryQuotaTracing: refund memory for redo log task",
+					zap.String("namespace", a.task.tableSink.changefeed.Namespace),
+					zap.String("changefeed", a.task.tableSink.changefeed.ID),
+					zap.Stringer("span", &a.task.span),
+					zap.Uint64("memory", refundMem))
+			}
+		}
+		if err := a.redoDMLManager.EmitRowChangedEvents(ctx, a.task.span, releaseMem,
+			a.events...); err != nil {
+			return errors.Trace(err)
+		}
+		a.events = a.events[:0]
```

Comment:
Should always re-create a buffer because `EmitRowChangedEvents` is asynchronous.